python/hidet/ir/tools/ir_dumper.py

```diff
@@ -0,0 +1,1630 @@
+
+        return DeclareStmt(var, value, is_static is not None, scope=scope)
+
+    def visit_fn_args(self, node):
+        return [self(v) for v in node]
+
+    def visit_function(self, node):
+        name = self(node[0])
+        self.cur_fn_name = name
+
+        args = self(node[1])
+        args = args if args is not None else []
+        return_type = self(node[2])
+        attrs = self(node[3])
+        body = self(node[4])
+
+        func_attrs = {}
+        if attrs in self.attributes:
+            func_attrs = self.attributes[attrs]
+        if 'func_kind' in func_attrs:
+            func_kind = func_attrs['func_kind']
+        elif 'cuda.grid_dim' in func_attrs or 'cuda.block_dim' in func_attrs:
+            if not all(name in func_attrs for name in ['cuda.grid_dim', 'cuda.block_dim']):
+                raise RuntimeError(
+                    name, 'CUDA kernel expects to have both attrs.cuda.grid_dim and attrs.cuda.block_dim.'
+                )
+            func_kind = 'cuda_kernel'
+        else:
+            func_kind = 'cuda_internal'
+
+        ir_fn = Function(name, args, body, return_type, func_kind, func_attrs)
+        func_var = Var(name, FuncType.from_func(ir_fn), name=name)
+        self.func_var_table[name] = func_var
+        self.functions[name] = ir_fn
+        return func_var
+
+    def visit_module(self, children):
+        mod_attrs = self.attributes[self(children[0])]
+
+        for stmt in children[1:]:
+            assert isinstance(stmt, Tree)
+            if stmt.data.value == 'function':
+                self(stmt)
+                self.cur_fn_name = None
+            elif stmt.data.value == 'declare_stmt':
+                self.visit(stmt)
+            elif stmt.data.value == 'attribute':
+                pass
+            else:
+                raise RuntimeError(f"Unknown stmt type {stmt.data.value}")
+
+        extern_funcs = {k: Var(None, self(v), name=k) for k, v in mod_attrs["extern_functions"].items()}
+        mod_attrs = {
+            "linking_libs": mod_attrs["linking_libs"],
+            "object_files": mod_attrs["object_files"],
+            "namespace": mod_attrs["namespace"],
+            "include_headers": mod_attrs["include_headers"],
+            "linking_dirs": mod_attrs["linking_dirs"],
+            "extern_functions": extern_funcs,
+        }
+
+        # maintain the same order of functions
+        # since python dicts preserve insertion order
+        functions = {}
+        for name in self.funcs:
+            functions[name] = self.functions[name]
+
+        return IRModule(functions, self.global_var_table, **mod_attrs)
+
+
+def diff_text(old: str, new: str):
+    from hidet.utils.py import red, green
+
+    old_lines = old.split('\n')
+    new_lines = new.split('\n')
+    for oline, nline in zip(old_lines, new_lines):
+        if oline != nline:
+            print(green(oline))
+            print(red(nline))
+        else:
+            print(oline)
+
+
+GRAMMAR = r"""
```

Comment:
Do we need to keep the grammar in both `hidet.lark` and in the source code?

It would be easier to keep the consistency of the grammar when it is only stored at one place. Maybe we can read the file and store it to `GRAMMAR` when we want to parse a dumped ir.

Reply:
It seems that reading files from within packages involves some fiddly work with paths. I found a solution that uses importlib_resources library, but only works for python >= 3.7, so I guess just including the grammar as a string is the simplest way.

Reply:
Yeah, let's go with adding the grammar directly in the source code.